packages/token/src/erc20/extensions/erc4626/erc4626.cairo

```diff
@@ -0,0 +1,709 @@
+/// "asset". The component leverages traits to configure fees, limits, and decimals.
+///
+/// CAUTION: In empty (or nearly empty) ERC-4626 vaults, deposits are at high risk of being stolen
+/// through frontrunning with a "donation" to the vault that inflates the price of a share. This is
+/// variously known as a donation or inflation attack and is essentially a problem of slippage.
+/// Vault deployers can protect against this attack by making an initial deposit of a non-trivial
+/// amount of the asset, such that price manipulation becomes infeasible. Withdrawals may similarly
+/// be affected by slippage. Users can protect against this attack as well as unexpected slippage in
+/// general by verifying the amount received is as expected, using a wrapper that performs these
+/// checks.
+///
+/// This implementation offers configurable virtual assets and shares to help developers mitigate
+/// that risk. `ImmutableConfig::DECIMALS_OFFSET` corresponds to an offset in the decimal
+/// representation between the underlying asset's decimals and vault decimals. This offset also
+/// determines the rate of virtual shares to virtual assets in the vault, which itself determines
+/// the initial exchange rate. While not fully preventing the attack, analysis shows that the
+/// default offset (0) makes it non-profitable even if an attacker is able to capture value from
+/// multiple user deposits, as a result of the value being captured by the virtual shares (out of
+/// the attacker's donation) matching the attacker's expected gains. With a larger offset, the
+/// attack becomes orders of magnitude more expensive than it is profitable.
+///
+/// The drawback of this approach is that the virtual shares do capture (a very small) part of the
+/// value being accrued to the vault. Also, if the vault experiences losses and users try to exit
+/// the vault, the virtual shares and assets will cause the first exiting user to experience reduced
+/// losses to the detriment to the last users who will experience bigger losses.
+#[starknet::component]
+pub mod ERC4626Component {
+    use core::num::traits::{Bounded, Pow, Zero};
+    use crate::erc20::ERC20Component;
+    use crate::erc20::ERC20Component::InternalImpl as ERC20InternalImpl;
+    use crate::erc20::extensions::erc4626::interface::IERC4626;
+    use crate::erc20::interface::{IERC20, IERC20Metadata};
+    use crate::erc20::interface::{IERC20Dispatcher, IERC20DispatcherTrait};
+    use openzeppelin_utils::math;
+    use openzeppelin_utils::math::Rounding;
+    use starknet::ContractAddress;
+    use starknet::storage::{StoragePointerReadAccess, StoragePointerWriteAccess};
+
+    // The default values are only used when the DefaultConfig
+    // is in scope in the implementing contract.
+    pub const DEFAULT_UNDERLYING_DECIMALS: u8 = 18;
+    pub const DEFAULT_DECIMALS_OFFSET: u8 = 0;
+
+    #[storage]
+    pub struct Storage {
+        ERC4626_asset: ContractAddress,
```

Comment:
```suggestion
        pub ERC4626_asset: ContractAddress,
```

As a general rule storage members should be accessible outside the component itself.